compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1082,6 +1082,20 @@ fn project_type<'cx, 'tcx>(
         return Ok(ProjectedTy::Progress(Progress::error(selcx.tcx())));
     }
 
+    // If the obligation contains any inference types or consts in associated
+    // type substs, then we don't assemble any candidates.
+    // This isn't really correct, but otherwise we can end up in a case where
+    // we constrain inference variables by selecting a single predicate, when
+    // we need to stay general. See issue #91762.
+    let (_, predicate_own_substs) = obligation.predicate.trait_ref_and_own_substs(selcx.tcx());
+    if predicate_own_substs.iter().any(|g| g.has_infer_types_or_consts()) {
+        return Ok(ProjectedTy::NoProgress(
```

Comment:
I think we want to treat this the same way we would treat an ambiguous set of candidates (i.e., `Err(ProjectionTyError::TooManyCandidates)`)